コメントをつけられるようにしました

naomichi-h opened the pull request:
Ruby on Railsのプラクティスサンプルアプリに、日報投稿機能とコメント機能を追加しました。
rubocop通過ずみです。

レビューのほど、よろしくお願いいたします。

naomichi-h committed:
変更を加える前にrubocopに通しました

naomichi-h committed:
日報投稿機能を作成しました

naomichi-h committed:
コメント機能を追加しました

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,70 @@
+    return unless @report.user_id == current_user.id
+
+    @report.destroy
+    respond_to do |format|
```
HTMLでのみレスポンスを返すのであれば `respond_to` は不要ですね。他のアクションも同様です。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,70 @@
+
+  # POST /reports or /reports.json
+  def create
+    @report = Report.new(report_params)
```
ここは `has_many` の関係を上手く活用すると、`user_id` の代入を自動化できます。

```ruby
@report = current_user.reports.build(report_params)
```

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,70 @@
+
+  # GET /reports or /reports.json
+  def index
+    @reports = Report.all
```
DBによっては順番がランダムとなる可能性があるので、明示的に `order` するようにしましょう。

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,26 @@
+      if @comment.save
+        format.html { redirect_to @commentable, notice: t('controllers.common.notice_create', name: Comment.model_name.human) }
+      else
+        format.html { redirect_to @commentable, notice: 'Comment was not successfully created.' }
```
※参考指摘なのでこのままでもOKです。

HTML側で `require: true` しているので、通常はここには来ないと思いますがここで `redirect_to` すると、`@comment.errors` の内容が消えてしまい、エラーの内容を画面に表示することができません。

もし、歓迎要件である「HTML5の必須チェックが効かない場合を考慮して、バリデーションエラー時のレスポンスも返せるようにする」を満たすのであれば、ここはセオリー通り `render` する必要があります。

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,26 @@
+
+  # Only allow a list of trusted parameters through.
+  def comment_params
+    params.require(:comment).permit(:content, :user_id)
```
`user_id` はフォームから入力しないので、ここでは不要ですね。

yuuu commented on app/models/comment.rb:
```diff
@@ -0,0 +1,6 @@
+
+class Comment < ApplicationRecord
+  belongs_to :user
+  belongs_to :commentable, polymorphic: true, dependent: :destroy
```
これだとコメントを削除したとき、BookやReportも一緒に消えてしまいそうですが問題ないでしょうか？

yuuu replied:
こちら、確かにコメントの削除はこのプラクティスにおける必須要件ではないのですが、修正いただきたいです🙏

というのも、これは「コメントの削除」というよりは「適切なモデル設計をする」という点で重要なためです。例えば、後からこのコードを読んだ開発者は、「このシステムではコメントを削除したら親リソースも削除する思想である」と誤った解釈をするかもしれません。

修正もオプションを削除するのみなので、そこまで大変ではないと思っています。もし何か懸念点などありましたらご相談ください 🙇

yuuu commented on app/views/comments/_index.html.erb:
```diff
@@ -0,0 +1,30 @@
+        <%= comment.user.name %>
+        <% end %>
+      </td>
+      <td><%= comment.created_at %></td>
```
日時を画面に表示するときは `l(comment.created_at)` とするとロケールに応じた見栄えにしてくれるのでおすすめです。

naomichi-h committed:
fix:レビューを受け、指摘箇所の修正を行いました。

不要なrespond_toを削除しました。
日報の新規作成のコードをリファクタリングしました。
日報の一覧表示の順番が明示的にidとなるようにしました。
日時を画面に表示する際、ロケールに応じた見栄えになるようにしました。
コメントモデルから、コメントを削除した際の不適切なオプションを削除しました。

naomichi-h commented:
指摘箇所の修正を行いました。
再度の確認のほど、よろしくお願いいたします。

※コメント削除など必須要件外の部分に関しては手をつけていません。

naomichi-h commented:
コメントモデルのオプションに関する修正を行いました。
よろしくお願いいたします。
